RFCs/0017-TestCase-FullyQualifiedName.md

```diff
@@ -1,49 +1,97 @@
+* `FullyQualifiedName` should be immutable in that the test store keeps the exact value that is provided by the test adapter and supplies it back to the test adapter for test execution. The test adapter should be able to rely on receiving the same FQN for execution that it supplied when performing discovery.
+* `FullyQualifiedName` should reference the semantic location of the method that defines the test. This means that the `FullyQualifiedName` references the method on the type where the test is discovered. In some situations, this could be different from the type where the method is defined. See later in this document for a more detailed discussion.
 
+## `FullyQualifiedName` specification
 
-The normalized form allows the test platform to reference groups of tests in a more general form without having to interpret the logic that the test adapter uses when constructing the fully qualified name. 
+We are standardizing the `FullyQualifiedName` property on a TestCase in the following way. The FQN will be parsed as a URI starting with the scheme value of `fqn`. 
+
+**Method Form**
+
+    fqn://clr/m/{type}/{method}
+
+`fqn://` - Any FQN that starts with the characters `fqn://` will be considered to follow the standardized format.
+
+`clr` – The first segment ("host" segment) specifies how to interpret the rest of the URI. The only valid value at the time of this writing is "clr" to reflect support for tests written in CLR managed languages. This portion can be expanded in the future to add support for non-managed languages such as Python, Javascript, C++, etc…
+
+`m | o` – `'m'` indicates a method form FQN, `'o'` indicates a custom test case identifier that the test explorer will not attempt to interpret.
+
+`{type}` - The `type` segment includes the fully specified type name in metadata format:
+
+    NamespaceA.NamespaceB.ClassName`1+InnerClass`2
+
+* The type name must be fully qualified (in the CLR sense), including its namespace. Any generic classes must also include an arity value using backtick notation (`# where # is the number of type arguments that the class requires).
+* Nested classes are appended with a '+' and must also include an arity if generic.
+* There should be no extra whitespace included in the segment
+
+`{method}` - The `method` segment includes the fully specified method including the method name and a list of its parameter types inside parentheses separated by commas.
+
+    MethodName`2(ParamTypeA,ParamTypeB,…)
+
+* If the method accepts no parameters, then the parentheses should be omitted.
+* If the method is generic, then an arity must be specified (in the same way as the type segment).
+* Insignificant whitespace should not be included in the segment.
+* The list of parameter types must be encoded in the same way that a type in the `{type}` segment of the FQN is encoded.
+
```

Comment:
Consider documenting how special names (overloaded operators, constructors, finalizers, etc should be named).

Reply:
I'm not sure how useful a test method could be if defined as an operator or finalizer, but constructors might have some potential. It makes sense to address it.

Reply:
I think Artur had done some investigation around operators etc. to see if we need to worry about them for source based discovery.Iirc at least some of the frameworks allow such cases with no errors.

At the very least we should be able to handle such tests without crashing (even if we can't fully adhere to the spec in such cases). It would be nice to add unit tests for this (I can look into this).